File: 프로그래머스/lv3/60059. 자물쇠와 열쇠/자물쇠와 열쇠.py
def rotate(arr):
    return list(zip(*arr[::-1]))

def get_new_lock(arr):
    new_lock=[[0]*3*len(arr) for _ in range(3*len(arr))]
    for i in range(len(arr)):
        for j in range(len(arr)):
            new_lock[i+len(arr)][j+len(arr)]= arr[i][j]
    return new_lock

def is_match(n,attached_lock):
    for i in range(n,2*n):
        for j in range(n,2*n):
            if attached_lock[i][j]!=1:
                return False
    return True
            
            
def attach(key,lock,x,y):
    for i in range(x,len(key)+x):
        for j in range(y,len(key)+y):
            lock[i][j]+=key[i-x][j-y]
    return lock
    
def check(n,m,key,lock):
    for i in range(n-m,2*n):
        for j in range(n-m,2*n):
            new_lock=get_new_lock(lock)
            attached_lock=attach(key,new_lock,i,j)
            if is_match(n,attached_lock):
                return True
    return False

def solution(key, lock):
    m,n=len(key),len(lock)
    answer = False
    if check(n,m,key,lock):
        return True
    
    for i in range(3):
        key=rotate(key)
        if check(n,m,key,lock):
            return True
    return answer

File: 프로그래머스/3/60059. 자물쇠와 열쇠/자물쇠와 열쇠.java
class Solution {
    
    private int groove=0;
    private int m, n;
    
    public boolean solution(int[][] key, int[][] lock) {
        boolean answer = false;
        
        m = key.length;
        n = lock.length;
        
        for(int i=0; i<n; i++){
            for(int j=0; j<n; j++){
                if(lock[i][j]==0) groove++;
            }
        }
        
        for(int dir=0; dir<4; dir++){
            key = rotate(key);
            
            for(int i=-n; i<=n; i++){
                for(int j=-n; j<=n; j++){
                    if(check(key, lock, i, j)){
                        return true;
                    }
                }
            }
        }
        
        return answer;
    }
    
    public boolean check(int[][] key, int[][] lock, int i , int j){
        int count=0;

        for(int k=0; k<m; k++){
            for(int l=0; l<m; l++){
                int nx = i+k;
                int ny = j+l;
                if(nx<0 || ny<0 || nx>=n || ny >=n) continue;
                if(key[k][l]==1 && lock[nx][ny]==1) return false;
                if(key[k][l]==1 && lock[nx][ny]==0){
                    count++;
                }
            }
        }
        if(count==groove){
            return true;
        }
        return false;
    }
    
    public int[][] rotate(int[][] key){
        int[][] rotated = new int[m][m];
        for(int i=0; i<m; i++){
            for(int j=m-1; j>=0; j--){
                rotated[i][m-j-1] = key[j][i];
            }
        }
        return rotated;
    }
}